services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,419 @@
+        return findTenantPromise.future()
+                .compose(tenantJsonResult -> Optional.ofNullable(tenantJsonResult)
+                        .map(tenantJson -> Future.succeededFuture(tenantJson.mapTo(TenantDto.class)))
+                        .orElse(Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                String.format("Tenant [%s] not found.",
+                                        findQuery.getString(RegistrationConstants.FIELD_PAYLOAD_TENANT_ID))))));
+    }
+
+    @Override
+    public Future<OperationResult<Id>> createTenant(final Optional<String> tenantId, final Tenant tenantObj,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(span);
+
+        final String tenantIdOrGenerated = tenantId.orElse(DeviceRegistryUtils.getUniqueIdentifier());
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantIdOrGenerated)
+                .compose(ok -> processCreateTenant(tenantIdOrGenerated, tenantObj, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<OperationResult<Id>> processCreateTenant(final String tenantId, final Tenant tenantObj,
+            final Span span) {
+
+        final TenantDto newTenantDto = new TenantDto(tenantId, tenantObj,
+                new Versioned<>(tenantObj).getVersion());
+
+        final JsonObject newTenantDtoJson = JsonObject.mapFrom(newTenantDto);
+        final Promise<String> createTenantPromise = Promise.promise();
+        mongoClient.insert(config.getCollectionName(), newTenantDtoJson, createTenantPromise);
+        return createTenantPromise.future()
+                .compose(tenantObjectIdResult -> {
+                    span.log("successfully created tenant.");
+                    return Future.succeededFuture(OperationResult.ok(
+                            HttpURLConnection.HTTP_CREATED,
+                            Id.of(tenantId),
+                            Optional.empty(),
+                            Optional.of(newTenantDto.getVersion())));
+                })
+                .recover(error -> {
+                    if (MongoDbDeviceRegistryUtils.isDuplicateKeyError(error)) {
+                        LOG.debug(String.format("The tenant [%s] already exists or an existing tenant uses a certificate authority with the same Subject DN", tenantId), error);
+                        TracingHelper.logError(span, "The tenant with the given identifier already exists or an existing tenant uses a certificate authority with the same Subject DN", error);
+                        return Future.succeededFuture(OperationResult.empty(HttpURLConnection.HTTP_CONFLICT));
+                    } else {
+                        LOG.error(String.format("Error adding Tenant [%s].", tenantId), error);
```

Comment:
no need to use `String.format` in a log statement

Reply:
Updated